public/service-worker.js

```diff
@@ -0,0 +1,90 @@
+const VERSION_CHECK_INTERVAL = 30 * 1000; // 30 sec
+const VERSION_URL = '/meta.json';
+
+// in sync with serviceWorker.ts
+const EventTypes = {
+  CLIENT_VERSION: 'CLIENT_VERSION',
+  NEW_VERSION_AVAILABLE: 'NEW_VERSION_AVAILABLE',
+};
+
+// holding the current client version
+let currentVersion = null;
+
+function isMajorOrMinorChanged(oldV, newV) {
+  if (!oldV || !newV) return false;
+  const [oMaj, oMin] = oldV.split('.');
+  const [nMaj, nMin] = newV.split('.');
+  return oMaj !== nMaj || oMin !== nMin;
+}
+
+function notifyClientsAboutUpdate(newVersion) {
+  self.clients.matchAll().then(clients => {
+    clients.forEach(client => {
+      client.postMessage({ type: EventTypes.NEW_VERSION_AVAILABLE, version: newVersion });
+    });
+  });
+}
+
+async function checkVersion() {
+  if (!currentVersion) {
+    return;
+  }
+
+  try {
+    const response = await fetch(VERSION_URL, { cache: 'no-store' });
+    const data = await response.json();
+    const newVersion = data.version;
+
+    if (newVersion !== currentVersion && isMajorOrMinorChanged(currentVersion, newVersion)) {
```

Comment:
Are we assuming we don't need to refresh on patch updates? Do we need to know the type of version bump or just check if the versions are different?